### PR TITLE
fix: from_string method of blob and bucket class

### LINKED ISSUE #286: Blob upload_from_string AttributeError when using client argument
#### Environment details

  - OS type and version: variant of Debian Linux
  - Python version: 3.8.5
  - pip version: 20.1.1
  - `google-cloud-storage` version: 1.31.2

#### Steps to reproduce

  1. create `Blob` object using `from_string` without a specified `client`
  2. use `Blob` `upload_from_string` with `client` argument

#### Code example

```python
from google.cloud import storage
from google.cloud.storage.blob import Blob 

def reproduce_error():
    storage_client = storage.Client()
    blob = Blob.from_string("gs://jdimatteo-gunk/example/upload_from_string_error/reproduce_error.txt")
    blob.upload_from_string("foo", client=storage_client)

def workaround_1():
    storage_client = storage.Client()
    blob = Blob.from_string(
            "gs://jdimatteo-gunk/example/upload_from_string_error/workaround_1.txt",
            client=storage_client)
    blob.upload_from_string("workaround_1\n")

def workaround_2():
    storage_client = storage.Client()
    blob = Blob.from_string("gs://jdimatteo-gunk/example/upload_from_string_error/workaround_2.txt")
    blob.bucket._client = storage_client
    blob.upload_from_string("workaround_2\n")

workaround_1()
workaround_2()
reproduce_error()
```

#### Stack trace
```
$ python example.py 
Traceback (most recent call last):
  File "example.py", line 24, in <module>
    reproduce_error()
  File "example.py", line 7, in reproduce_error
    blob.upload_from_string("foo", client=storage_client)
  File "/usr/local/google/home/jdimatteo/gunk/upload_from_string_error_example/venv/lib/python3.8/site-packages/google/cloud/storage/blob.py", line 2440, in upload_from_string
    self.upload_from_file(
  File "/usr/local/google/home/jdimatteo/gunk/upload_from_string_error_example/venv/lib/python3.8/site-packages/google/cloud/storage/blob.py", line 2223, in upload_from_file
    created_json = self._do_upload(
  File "/usr/local/google/home/jdimatteo/gunk/upload_from_string_error_example/venv/lib/python3.8/site-packages/google/cloud/storage/blob.py", line 2055, in _do_upload
    response = self._do_multipart_upload(
  File "/usr/local/google/home/jdimatteo/gunk/upload_from_string_error_example/venv/lib/python3.8/site-packages/google/cloud/storage/blob.py", line 1614, in _do_multipart_upload
    hostname=self.client._connection.API_BASE_URL, bucket_path=self.bucket.path
AttributeError: 'NoneType' object has no attribute '_connection'
$ gsutil cat gs://jdimatteo-gunk/example/upload_from_string_error/workaround_1.txt
workaround_1
$ gsutil cat gs://jdimatteo-gunk/example/upload_from_string_error/workaround_2.txt
workaround_2
```

### PR DESCRIPTION
**Pull request by HemangChothani**
Fixes #286

### CODE REVIEW

File: google/cloud/storage/blob.py [hunk 1]
```diff
@@ -330,15 +330,15 @@ def public_url(self):
         )
 
     @classmethod
-    def from_string(cls, uri, client=None):
+    def from_string(cls, uri, client):
```
Comment (tseaver) [inline google/cloud/storage/blob.py:333]: Making a formerly optional argument required is a breaking change.

**Review by tseaver — CHANGES_REQUESTED** [review]
Rather than changing the signature of `Blob.from_string`, the fix should be to change `Blob._do_multipart_upload` (and likewise `Blob._initiate_resumable_upload`) to use the passed-in client. The caller (`Blob._do_upload`) should likewise call `self._require_client` to use its own `_client` attribute as a fallback for when no explicit `client` is passed in.

File: tests/unit/test_blob.py [hunk 2]
```diff
@@ -1982,7 +1983,7 @@ def test__do_multipart_upload_bad_size(self):
 
         with self.assertRaises(ValueError) as exc_info:
             blob._do_multipart_upload(
-                None, stream, None, size, None, None, None, None, None, None
+                client, stream, None, size, None, None, None, None, None, None
             )
```
Comment (tseaver) [inline tests/unit/test_blob.py:1987]: Given that the passed-in `client` is not used by `Blob._do_multipart_upload` before the `ValueError` is raised, this change is unnecessary / distracting.

We do need extra test cases that verify that the passed-in `client` is used in preference to `self._client` in both `Blob._do_multipart_upload` and `Blob._initiate_resumable_upload`.

**Review by tseaver — CHANGES_REQUESTED** [review]

**Comment by HemangChothani** [comment]
@tseaver PTAL!

**Review by tseaver — APPROVED** [review]

**Pull request merged**